metals/src/main/scala/scala/meta/internal/metals/Ammonite.scala

```diff
@@ -0,0 +1,390 @@
+        importedBuild0
+      )
+      _ = {
+        lastImportedBuild0 = build0
+      }
+      _ <- indexWorkspace()
+      // _ = checkRunningBloopVersion(build.version)
+      _ <- Future.sequence[Unit, List](
+        compilations()
+          .cascadeCompileFiles(buffers.open.toSeq) ::
+          compilers().load(
+            buffers.open.toSeq.filter(!_.isAmmoniteScript)
+          ) ::
+          Nil
+      )
+    } yield BuildChange.Reconnected
+  }
+
+  private def disconnectOldBuildServer(): Future[Unit] = {
+    if (buildServer0.isDefined)
+      scribe.info("disconnected: ammonite build server")
+    buildServer0 match {
+      case None => Future.successful(())
+      case Some(value) =>
+        // remove from cancellable too?
+        buildServer0 = None
+        lastImportedBuild0 = ImportedBuild.empty
+        cancelables.remove(value)
+        diagnostics().resetAmmoniteScripts()
+        value.shutdown()
+    }
+  }
+
+  def maybeImport(path: AbsolutePath): Unit =
+    if (path.isAmmoniteScript && !loaded(path))
+      languageClient()
+        .showMessageRequest(Messages.ImportAmmoniteScript.params())
+        .asScala
+        .onComplete {
+          case Failure(e) => scribe.warn("Error requesting Ammonite import", e)
+          case Success(resp) =>
+            resp.getTitle match {
+              case Messages.ImportAmmoniteScript.doImport =>
+                start(Some(path)).asScala.onComplete {
+                  case Failure(e) =>
+                    scribe.warn(s"Error importing Ammonite script $path", e)
+                  case Success(_) =>
+                }
+              case _ =>
+            }
+        }
+
+  def start(): CompletableFuture[Object] =
+    start(None)
+
+  def start(doc: Option[AbsolutePath]): CompletableFuture[Object] = {
+
+    disconnectOldBuildServer().onComplete {
+      case Failure(e) =>
+        scribe.warn("Error disconnecting old Ammonite build server", e)
+      case Success(()) =>
+    }
+
+    // TODO Look at buffers.open rather than focusedDocument?
+    val commandScriptOpt = doc.orElse(focusedDocument()) match {
+      case None =>
+        val msg = "No Ammonite script is opened"
+        scribe.error(msg)
+        Left(new Exception(msg))
+      case Some(path) if path.isAmmoniteScript =>
+        val it = path.toInputFromBuffers(buffers).value.linesIterator
+        val versionsOpt = VersionsOption.fromScript(it)
+        val versions = versionsOpt
+          .orElse(lastImportVersions)
+          .getOrElse(
+            // TODO Warn if any of these is used
+            AmmVersions
+              .default()
+              // temporary, remove once Ammonite is available for 2.12.11
+              // .withScalaVersion(BuildInfo.scala212)
+              .withScalaVersion("2.12.10")
+              // do not merge, remove once lihaoyi/Ammonite#1053 is merged
+              .withAmmoniteVersion("2.0.5-bsp-3")
+          )
+        AmmoniteFetcher(versions)
```

Comment:
Maybe let's move logic to fetch Ammonite to a separate method.

Reply:
Done, it's better this way, yes.